File: packages/installer-server/src/lib/steps/solana.ts
import { execa } from 'execa';
import { stageScript } from '../downloader.js';
import type { InstallerStep } from './index.js';

const ANZA_INSTALL_URL = 'https://release.anza.xyz/stable/install';

export const solanaStep: InstallerStep = {
  id: 'solana-cli',
  label: 'Install Solana CLI (Anza release channel)',
  shouldRun: ({ submission }) => submission.optionalFeatures.solana.cli,
  async run({ bus, platform }) {
    if (platform === 'win32') {
      bus.log(
        solanaStep.id,
        'warn',
        'Solana CLI on native Windows requires WSL; skipping. Re-run inside WSL2 to install.',
      );
      return;
    }

    bus.log(solanaStep.id, 'info', `Downloading installer from ${ANZA_INSTALL_URL}...`);
    const staged = await stageScript(ANZA_INSTALL_URL);
    bus.log(solanaStep.id, 'info', `Staged at ${staged.path} (sha256=${staged.sha256}, ${staged.size} bytes).`);

    const sub = execa('sh', [staged.path], { all: true });
    sub.all?.on('data', (chunk: Buffer) => {
      const text = chunk.toString().trimEnd();
      if (text) bus.log(solanaStep.id, 'info', text);
    });
    await sub;

    bus.log(solanaStep.id, 'info', 'Add ~/.local/share/solana/install/active_release/bin to PATH.');
  },
};

File: internal/installer/steps/solana.go
package steps

import (
	"context"
	"fmt"
	"runtime"

	"github.com/SolsticeSauer/solsclaw_beta/internal/installer"
)

const anzaInstallURL = "https://release.anza.xyz/stable/install"

type Solana struct{}

func (Solana) ID() string    { return "solana-cli" }
func (Solana) Label() string { return "Install Solana CLI (Anza release channel)" }

func (Solana) ShouldRun(sc installer.StepContext) bool {
	// Solana CLI inside a transient container has no good "where does the
	// keypair live" answer, so we leave it as a host-only step. Users on
	// Docker who need Solana can either add it to the Dockerfile or run
	// it on the host alongside.
	if sc.Submission.InstallMode == installer.ModeDocker {
		return false
	}
	return sc.Submission.OptionalFeatures.Solana.CLI
}

func (s Solana) Run(ctx context.Context, sc installer.StepContext) error {
	if runtime.GOOS == "windows" {
		sc.Bus.Log(s.ID(), "warn", "Solana CLI on native Windows requires WSL; skipping. Re-run inside WSL2 to install.")
		return nil
	}

	sc.Bus.Log(s.ID(), "info", "Downloading installer from "+anzaInstallURL)
	staged, err := stage(anzaInstallURL)
	if err != nil {
		return err
	}
	sc.Bus.Log(s.ID(), "info", fmt.Sprintf("Staged at %s (sha256=%s, %d bytes).", staged.Path, staged.SHA256, staged.Size))

	if err := streamCommand(ctx, sc, s.ID(), "sh", staged.Path); err != nil {
		return err
	}
	sc.Bus.Log(s.ID(), "info", "Add ~/.local/share/solana/install/active_release/bin to PATH.")
	return nil
}
